src/main/scala/org/ergoplatform/nodeView/ErgoModifiersCache.scala

```diff
@@ -0,0 +1,61 @@
+package org.ergoplatform.nodeView
+
+import org.ergoplatform.modifiers.ErgoPersistentModifier
+import org.ergoplatform.modifiers.history.Header
+import org.ergoplatform.nodeView.history.ErgoHistory
+import scorex.core.{LRUCache, ModifiersCache}
+import scorex.core.utils.ScorexLogging
+import scorex.core.validation.RecoverableModifierError
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+import scala.util.{Failure, Random, Success}
+
+class ErgoModifiersCache(override val maxSize: Int)
+  extends ModifiersCache[ErgoPersistentModifier, ErgoHistory]
+      with LRUCache[ErgoPersistentModifier, ErgoHistory]
+      with ScorexLogging  {
+
+  protected override def onPut(key: K): Unit =
+    super.onPut(key)
+
+  protected override def onRemove(key: K, rememberKey: Boolean): Unit =
+    super.onRemove(key, rememberKey)
+
+  override def findCandidateKey(history: ErgoHistory): Option[K] = {
+    val headersHeight = history.headersHeight
+
+    val maxIterations = 10
+
+    @tailrec
+    def cacheIteration(called: Int): Option[K] = {
+      val expectedFullBlockParts = history
+        .headerIdsAtHeight(history.fullBlockHeight + 1)
```

Comment:
What if we have a fork? E.g current chain is `A1  ->  A2` (height 2)  and we have blocks B2 and B3 in cache. Looks like we'll never try to apply block B2, because it's height is equals to history.fullBlockHeight. 
A1  ->  A2
      \-> B2-> B3